analyzers/src/SonarAnalyzer.Common/Rules/SecurityPInvokeMethodShouldNotBeCalledBase.cs

```diff
@@ -0,0 +1,83 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class SecurityPInvokeMethodShouldNotBeCalledBase<TSyntaxKind, TInvocationExpressionSyntax, TIdentifierNameSyntax> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
+        where TInvocationExpressionSyntax : SyntaxNode
+        where TIdentifierNameSyntax : SyntaxNode
+    {
+        protected const string DiagnosticId = "S3884";
+        protected const string MessageFormat = "Refactor the code to remove this use of '{0}'.";
+        protected const string InteropDllName = "ole32.dll";
+
+        protected abstract TSyntaxKind SyntaxKind { get; }
+        protected abstract ILanguageFacade Language { get; }
+
+        protected abstract SyntaxNode Expression(TInvocationExpressionSyntax invocationExpression);
+        protected abstract SyntaxToken Identifier(TIdentifierNameSyntax identifierName);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected DiagnosticDescriptor Rule { get; }
+
+        protected ISet<string> InvalidMethods { get; } = new HashSet<string>
+        {
+            "CoSetProxyBlanket",
+            "CoInitializeSecurity"
+        };
+
+        protected SecurityPInvokeMethodShouldNotBeCalledBase(System.Resources.ResourceManager rspecResources) =>
+            Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, CheckForIssue, SyntaxKind);
+
+        protected virtual bool IsImportFromInteropDll(ISymbol symbol) =>
+            symbol.GetAttributes(KnownType.System_Runtime_InteropServices_DllImportAttribute).FirstOrDefault() is AttributeData attributeData
+            && attributeData.ConstructorArguments.Any(x => x.Value.Equals(InteropDllName));
+
+        protected virtual string GetMethodName(ISymbol symbol, SyntaxToken syntaxToken) =>
+            syntaxToken.ValueText;
+
+        private void CheckForIssue(SyntaxNodeAnalysisContext analysisContext)
+        {
+            if (analysisContext.Node is TInvocationExpressionSyntax invocation
+                && Expression(invocation) is TIdentifierNameSyntax directMethodCall
+                && analysisContext.SemanticModel.GetSymbolInfo(directMethodCall) is SymbolInfo methodCallSymbol
+                && methodCallSymbol.Symbol != null
+                && InvalidMethods.Contains(GetMethodName(methodCallSymbol.Symbol, Identifier(directMethodCall)))
```

Comment:
This name check was originally before symbol retrieving for C#. That's a major performance optimization, because it's rare to have methods named like this. Currently, we'd get symbol for each invocation for C# and that's expensive.

I think it's worth the trouble to have this optimization in place for C# (and re-validate the name for 2nd time after symbol) and to not have it for VB, as there's no other way around it (only the current symbol based check will kick in)